src/rt/sched/cown.h

```diff
@@ -953,54 +917,48 @@ namespace verona::rt
       }
     }
 
-    /// Update backpressure status based on the occurrence of a token message.
-    /// Return true if the current message is a token.
-    inline bool check_message_token(Alloc* alloc, MessageBody* curr)
+    /// Return true if this cown's message queue is growing faster than the
+    /// messages are being processed. This is based on an approximation for the
+    /// length of this cown's message queue.
+    inline bool overloaded() const
     {
-      auto stat = status.load(std::memory_order_acquire);
+      static constexpr size_t overload_rc = 800;
+      const auto rc = get_header().rc.load(std::memory_order_relaxed);
+      assert(rc != 0);
       yield();
+      return rc >= overload_rc;
+    }
+
+    /// Update priority based on the occurrence of a token message, or replace
+    /// the token if it is not in the queue. Return true if the current message
+    /// is a token.
+    inline bool check_message_token(Alloc* alloc, MessageBody* curr)
+    {
       if (curr == nullptr)
       {
         Systematic::cout() << "Reached message token on cown " << this
                            << std::endl;
-        assert(stat.has_token());
-        stat.set_has_token(false);
-        status.store(stat, std::memory_order_release);
+        assert(has_token);
+        has_token = false;
+        if (overloaded())
```

Comment:
Could the token message track the RC last time we sent the token message, so cowns with high in degree can not be considered overloaded, i.e. if the RC is not growing, then it don't consider it overloaded?

Reply:
Also, does it make sense to scan the message queue for its length when you observe a high RC?  I.e. most of the time, we know it is not a problem due to the RC, and only scan if it is a problem.

Reply:
I'll look into this later today.

Reply:
Checking `next->overloaded()` in `fast_send` means that we can't check the actual queue length when needed. So instead, it should correct for the in-degree using a value updated whenever a cown finds its token message. The in-degree correction could be compressed into 8 bits using something like the snmalloc sizeclass, but that still raises the problem of having to make space for 8 bits. Does that make sense, or do you have something in mind that takes less space to correct for the in-degree?